js/webserver.js

```diff
@@ -104,62 +143,94 @@ WebServer.prototype = {
+        let httpsServer;
+        t._loadHttpsKeyData();
+        while (!listening) {
+          try {
+            httpsServer = https.createServer(t.httpsOptions, app);
+            t._setErrorLogger(httpsServer, 'HTTPS', ipAddress, port);
+            t.httpsServers.push(httpsServer);
+            t.expressWsHttps.push(expressWs(app, httpsServer, {maxPayload: 50000}));
+            listening = true;
+          } catch (e) {
+            if (e.message == 'mac verify failure') {
+              const r = reader();
+              try {
+                t.httpsOptions.passphrase = yield reader.readPassword(
+                  'HTTPS key or PFX decryption failure. Please enter passphrase: ');
+              } finally {
+                r.close();
+              }
+            } else {
+              throw e;
             }
-          } else {
-            throw e;
           }
         }
+        t.callListen(httpsServer, 'https', 'HTTPS', ipAddress, port);
+      };
+
+     
+      if (this.config.https.ipAddresses) {
+        let httpsIps = sanitizeIps(this.config.https.ipAddresses);
+        let len = httpsIps.size;
+        if (len > 0) {
+          const iterator = httpsIps.entries();
+          for (let ipAddress of iterator) {
+            yield* makeHttpsServer(ipAddress[0], this.config.https.port);
+          }
+        } else {
+          makeHttpsServer('0.0.0.0', this.config.https.port);
+        }
+      } else {
+        makeHttpsServer('0.0.0.0', this.config.https.port);
       }
-      this.callListen(this.httpsServer, 'https', 'HTTPS');
     }
     if (this.config.http && this.config.http.port) {
-      this.httpServer = http.createServer(app);
-      this.expressWsHttp = expressWs(app, this.httpServer);
-      this.callListen(this.httpServer, 'http', 'HTTP');
+      let makeHttpServer = function(ipAddress, port) {
+        let httpServer = http.createServer(app);
+        t._setErrorLogger(httpServer, 'HTTP', ipAddress, port);
+        t.httpServers.push(httpServer);
+        t.expressWsHttp.push(expressWs(app, httpServer));
+        t.callListen(httpServer, 'http', 'HTTP', ipAddress, port);
+      }
+      
+      if (this.config.http.ipAddresses) {
+        let httpIps = sanitizeIps(this.config.http.ipAddresses);
+        if (httpIps.size > 0) {
+          httpIps.forEach((ipAddress)=> {
+            makeHttpServer(ipAddress, this.config.http.port);
+          });
+        } else {
+          makeHttpServer('0.0.0.0', this.config.http.port);
+        }
+      } else {
+        makeHttpServer('0.0.0.0', this.config.http.port);
+      }
     }
   }),
 
-  callListen(methodServer, methodName, methodNameForLogging) {
-    var methodConfig = this.config[methodName];
-    var addressForLogging = methodConfig.hostname ? methodConfig.hostname : "*";
-    addressForLogging += ":" + methodConfig.port;
-
+  callListen(methodServer, methodName, methodNameForLogging, ipAddress, port) {
+    var addressForLogging = `${ipAddress}:${port}`;
```

Comment:
The use of `var` vs `let` is inconsistent in this PR. What I think makes sense is using `const` by default  and that will be like 90% of cases. Then, once you need a modifiable variable, you use `let` (so once a reader sees a `let` they know they should expect the value to be assigned more than once). And you typically don't use `var` at all, unless you want the weird 'hoisting' behaviour for some reason.

Reply:
Right; sometimes this gets missed due to habit. I'm fixing this.